README.md
<img align=right width=33% src="assets/logo.svg"/>

# Screech
Compact test-driven development assistance tool for C.

---

This compact library has been used for swiftly writing tests when working on C projects using a test-driven development methodology.

It is intended to be compact and simple to write tests in.

---

#### To compile the library and self-test:

Generate makefiles with [Premake 5](https://github.com/premake/premake-core), e.g.

`$ premake5 gmake`

Build with the desired tool, e.g.

`$ make`

No auto-installation script is available at the moment.


CHANGELOG.md
## Screech 0.1.4 :|: 2020-07-26
* First public release.
* Added graphic assets for decoration.


test/screech.test.c
/*
 *  Name:       screech.test.c
 *  Purpose:    Screech self-testing test.
 *  Updated:    2020-07-26
 *  Author:     SmiVan
 */

#include <stdlib.h>
#include <stdio.h>
#include "screech.h"

int example_test(void) {
    return 1 == 1;
}

int bad_test(void) {
    return 1 == 0;
}

int test_once(void) {
    return screech_test("Testing Once", 1,
        (test){.function = example_test, .description = "Example Test 1.1"}
    );
}

int test_thrice(void) {
    return screech_test("Testing Thrice", 3,
        (test){.function = example_test, .description = "Example Test 2.1"},
        (test){.function = example_test, .description = "Example Test 2.2"},
        (test){.function = example_test, .description = "Example Test 2.3"}
    );
}

int test_failed(void) {
    return !screech_test("Testing Failing", 3,
        (test){.function = example_test, .description = "Example Test 3.1"},
        (test){.function = example_test, .description = "Example Test 3.2"},
        (test){.function = bad_test, .description = "Bad Test"}
    );
}

int main(void) {
    screech_test("Screech Recursive Test", 3,
        (test){.function = test_once, .description = "Test Once"},
        (test){.function = test_thrice, .description = "Test Thrice"},
        (test){.function = test_failed, .description = "Test Failed"}
    );
    return 0;
}


source/screech.c
/*
 *  Name:       screech.c
 *  Purpose:    Test-driven development assistance tool for C.
 *  Updated:    2020-07-26
 *  Author:     SmiVan
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include "screech.h"

// The following is a list of tests to perform.
typedef struct test_suite {
    test* tests;
    const char* description;
    int count;
} test_suite;

typedef struct scoreboard {
    int testsPassed;
    int testsFailed;
} scoreboard;

void run_test(int (*test_function)(void), int test_number, const char* test_description, scoreboard* score) {

    if (test_function()) {
        // Test passed
        score->testsPassed += 1;
        fprintf(stderr, TEST_RESULT_FORMAT, test_number, test_description, TEST_PASSED_MESSAGE);
    } else {
        // Test failed
        score->testsFailed += 1;
        fprintf(stderr, TEST_RESULT_FORMAT, test_number, test_description, TEST_FAILED_MESSAGE);
    }
}

int run_suite(test_suite* suite) {
    fprintf(stderr, "\nTesting \"%s\"...\n", suite->description);

    // SETUP //
    scoreboard* score = (scoreboard*) malloc(sizeof(scoreboard));
    score->testsPassed = 0;
    score->testsFailed = 0;

    // TESTS  BEGIN //
    int test_counter = 0;
    test* current_test = suite->tests;

    while (test_counter < suite->count) {

        //increase test counter
        test_counter++;

        //run the test
        run_test(current_test->function, test_counter, current_test->description, score);

        //shift cursor to next test
        current_test = current_test + 1;
    }
    // TESTS   END  //

    printf("%i/%i tests passed.\n", score->testsPassed, score->testsPassed + score->testsFailed);

    if(score->testsFailed == 0) {
        printf("Working as intended.\n");
    }

    int failCount = score->testsFailed;

    free(score);

    return failCount;
}

// make test
test* make_test(test_func function, char* description) {
    test* this_test = (test*) malloc(sizeof(test));
    this_test->function = function;
    this_test->description = description;

    return this_test;
};

// free test
int free_test(test* test) {
    free(test);
    return 0;
};

// free test suite
int free_test_suite(test_suite* suite) {
    free(suite->tests);
    free(suite);
    return 0;
};

// make test suite, test using that suite and then free it
int screech_test(const char* description, int count, ...) {

    // MAKING TEST SUITE
    // loosely based on https://www.eskimo.com/~scs/cclass/int/sx11b.html

    test* test_list = (test*) calloc(count, sizeof(test));

    va_list argp;

    va_start(argp, count); // reads varargs *after* count argument

    // it's impossible to know how many varargs there are without count
    for(int i = 0; i < count; i++) {
        test_list[i] = va_arg(argp, test);
    }

    va_end(argp);

    test_suite* suite = (test_suite*) malloc(sizeof(test_suite));
    suite->tests = test_list;
    suite->description = description;
    suite->count = count;

    // TEST SUITE COMPLETE

    int fail_count = run_suite(suite);

    free_test_suite(suite);

    return !fail_count;
};


include/screech.h
/*
 *  Name:       screech.h
 *  Purpose:    Test-driven development assistance tool for C.
 *  Updated:    2020-07-26
 *  Author:     SmiVan
 */

#ifndef SCREECH_H_
#define SCREECH_H_

#ifdef __cplusplus
extern "C" {
#endif

#define TEST_RESULT_FORMAT "%d. %s - %s"
#define TEST_PASSED_MESSAGE "passed\n"
#define TEST_FAILED_MESSAGE "failed\n"

// The following defines a type for the test function pointer,
// that takes in nothing and returns int.
typedef int (*test_func)(void);

// The following is a struct with a test function and it's description.
typedef struct test {
    test_func function;
    const char* description;
} test;

/* Screech easy testing main function.
 *
 * Example usage:
 *
 * return screech_test("Testing Thrice", 3,
 *     (test){.function = example_test, .description = "Example Test 2.1"},
 *     (test){.function = example_test, .description = "Example Test 2.2"},
 *     (test){.function = example_test, .description = "Example Test 2.3"}
 * );
 *
 */
int screech_test(const char* description, int test_count, ...);

#ifdef __cplusplus
}
#endif

#endif
